relay-general/src/pii/event_paths.rs

```diff
@@ -0,0 +1,213 @@
+/// completed.
+///
+/// The main value in autocompletion is that we can complete `$http.headers.Authorization` as soon
+/// as the user types `Auth`.
+///
+/// XXX: This function should not have to take an event by value, we only do that
+/// due to restrictions on the Processor trait that we internally use to traverse the event.
+pub fn selectors_from_event(mut event: Annotated<Event>) -> BTreeSet<String> {
+    let mut processor = EventPathsProcessor {
+        paths: BTreeSet::new(),
+    };
+
+    process_value(&mut event, &mut processor, ProcessingState::root())
+        .expect("This processor is supposed to be infallible");
+
+    processor.paths
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_empty() {
+        // Test that an event without PII will generate empty list.
+        let event = Annotated::<Event>::from_json(r#"{"logentry": {"message": "hi"}}"#).unwrap();
+
+        let selectors = selectors_from_event(event);
+        insta::assert_yaml_snapshot!(selectors, @r###"
+        ---
+        []
+        "###);
+    }
+
+    #[test]
+    fn test_full() {
+        let event = Annotated::<Event>::from_json(
+            r##"
+            {
+              "message": "hi",
+              "exception": {
+                "values": [
+                  {
+                    "type": "ZeroDivisionError",
+                    "value": "Divided by zero",
+                    "stacktrace": {
+                      "frames": [
+                        {
+                          "abs_path": "foo/bar/baz",
+                          "filename": "baz",
+                          "vars": {
+                            "foo": "bar"
+                          }
+                        }
+                      ]
+                    }
+                  },
+                  {
+                    "type": "BrokenException",
+                    "value": "Something failed",
+                    "stacktrace": {
+                      "frames": [
+                        {
+                          "vars": {
+                            "bam": "bar"
+                          }
+                        }
+                      ]
+                    }
+                  }
+                ]
+              },
+              "extra": {
+                "My Custom Value": "123"
+              },
+              "request": {
+                "headers": {
+                  "Authorization": "not really"
+                }
+              }
+            }
+            "##,
+        )
+        .unwrap();
+
+        let selectors = selectors_from_event(event);
+        insta::assert_yaml_snapshot!(selectors, @r###"
+        ---
+        - $frame.abs_path
+        - $frame.filename
+        - $frame.vars
+        - $frame.vars.bam
+        - $frame.vars.foo
+        - $http.headers
+        - $http.headers.Authorization
+        - $message
+        - $string
+        - extra
+        - "extra.'My Custom Value'"
```

Comment:
I would not autocomplete into these custom fields for future-proofing. 

Especially in case we want to move to schema-based autocompletion in the future, we could no longer sustain such a feature. Just completing on the known schema is already quite a benefit, and if customers are going in to configure a specific selector for one of their extras/vars, then they probably know what they want anyway (and it's also likely that it doesn't show up under the last 10 anyway).

Reply:
From a product perspective I want to sustain such a feature independently of what internal changes we do. If you say schema-based autocompletion is conflicting with this, then I'd rather never have schema-based autocompletion.

Reply:
BTW I don't think schema-based autocompletion is conflicting with this. I can imagine that we do both depending on whether the user came from an event.

Reply:
I fundamentally and strongly disagree with this product decision, but this is your call to make 👍

Reply:
we synced via google hangouts and I showed @jan-auer how the UI looks like, we're now on the same page at least for the short/mid-term